pkg/kubelet/dockershim/docker_container_windows.go

```diff
@@ -0,0 +1,224 @@
+// it copies its contents to a unique registry value, and sets a SecurityOpt on the config pointing to that registry value.
+// We use registry values instead of files since their location cannot change - as opposed to credential spec files,
+// whose location could potentially change down the line, or even be unknown (eg if docker is not installed on the
+// C: drive)
+// When docker supports passing a credential spec's contents directly, we should switch to using that
+// as it will avoid cluttering the registry.
+func applyGMSAConfig(config *runtimeapi.ContainerConfig, createConfig *dockertypes.ContainerCreateConfig, cleanupInfo *containerCreationCleanupInfo) error {
+	credSpec := config.Annotations[kuberuntime.GMSASpecContainerAnnotationKey]
+	if credSpec == "" {
+		return nil
+	}
+
+	valueName, err := copyGMSACredSpecToRegistryValue(credSpec)
+	if err != nil {
+		return err
+	}
+
+	if createConfig.HostConfig == nil {
+		createConfig.HostConfig = &dockercontainer.HostConfig{}
+	}
+
+	createConfig.HostConfig.SecurityOpt = append(createConfig.HostConfig.SecurityOpt, "credentialspec=registry://"+valueName)
+	cleanupInfo.gMSARegistryValueName = valueName
+
+	return nil
+}
+
+const (
+	// same as https://github.com/moby/moby/blob/93d994e29c9cc8d81f1b0477e28d705fa7e2cd72/daemon/oci_windows.go#L23
+	credentialSpecRegistryLocation = `SOFTWARE\Microsoft\Windows NT\CurrentVersion\Virtualization\Containers\CredentialSpecs`
+	// the prefix for the registry values we write GMSA cred specs to
+	gMSARegistryValueNamePrefix = "k8s-cred-spec-"
+	// the number of random bytes to generate suffixes for registry value names
+	gMSARegistryValueNameSuffixRandomBytes = 40
+)
+
+// registryKey is an interface wrapper around `registry.Key`,
+// listing only the methods we care about here.
+// It's mainly useful to easily allow mocking the registry in tests.
+type registryKey interface {
+	SetStringValue(name, value string) error
+	DeleteValue(name string) error
+	ReadValueNames(n int) ([]string, error)
+	Close() error
+}
+
+var registryCreateKeyFunc = func(baseKey registry.Key, path string, access uint32) (registryKey, bool, error) {
+	return registry.CreateKey(baseKey, path, access)
+}
+
+// randomReader is only meant to ever be overridden for testing purposes,
+// same idea as for `registryKey` above
+var randomReader = rand.Reader
+
+// gMSARegistryValueNamesRegex is the regex used to detect gMSA cred spec
+// registry values in `removeAllGMSARegistryValues` below.
+var gMSARegistryValueNamesRegex = regexp.MustCompile(fmt.Sprintf("^%s[0-9a-f]{%d}$", gMSARegistryValueNamePrefix, 2*gMSARegistryValueNameSuffixRandomBytes))
+
+// copyGMSACredSpecToRegistryKey copies the credential specs to a unique registry value, and returns its name.
+func copyGMSACredSpecToRegistryValue(credSpec string) (string, error) {
+	valueName, err := gMSARegistryValueName()
+	if err != nil {
+		return "", err
+	}
+
+	// write to the registry
```

Comment:
I wasn't expecting registry modification as part of this... the original design indicated a file-based mechanisms would be used:

> Docker (to be supported in Alpha): will receive the path to a file created on the node's file system under C:\ProgramData\docker\CredentialSpecs\ and populated by dockershim with the credential spec JSON. Docker will read the contents of the credential spec file and pass it to Windows Host Compute Service (HCS) when creating and starting a container in Windows.

https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md#changes-in-dockershim:

> Changes in Dockershim
> During Alpha, dockerService.CreateContainer function will be enhanced (under a feature flag: WindowsGMSA) to create a temporary file with a unique name on the host file system under path C:\ProgramData\docker\CredentialSpecs\. This file will be populated with the contents of container.alpha.windows.kubernetes.io/gmsa-credential-spec annotation in CreateContainerRequest.ContainerConfig. Beta onwards, dockerService.CreateContainer (under a feature flag: WindowsGMSA) will use the contents of WindowsContainerSecurityContext.CredentialSpec to populate the file.

Did that approach not work?

Reply:
Sadly there is no robust way to know where Docker expects its cred spec files. As far as I could tell, Docker does not provide any reliable way to know where it's been installed, and that's where it expects cred spec files; for that reason, registry keys seemed like more viable option.

That being said, as I mentioned in https://github.com/kubernetes/kubernetes/pull/73726/files#r257464340, Docker should soon support passing base64-encoded cred spec directly, thus removing the need for all this. This is just needed in the meantime, and until k8s' stops supporting the last version of Docker that won't have that feature.

Reply:
@wk8 i am ok with this change since we will update it in the future (passing cred spec directly). However, docker on windows does tell you where it is installed in this registry key.

HKEY_LOCAL_MACHINE\SOFTWARE\Docker Inc.\Docker\1.0
Loop for the AppPath/BinPath Strings.

Reply:
@michmike : interesting, I don't have these reg keys on my Win boxes. Installed with
```powershell
Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
Install-Package -Name Docker -ProviderName DockerMsftProvider
```

Reply:
And for the record, the moby PR for this is out: https://github.com/moby/moby/pull/38777 ; will add that as a comment in this one.

Reply:
> @michmike : interesting, I don't have these reg keys on my Win boxes - `HKEY_LOCAL_MACHINE\SOFTWARE\Docker Inc.` just doesn't exist, and grepping for `docker` (case -insensitive) throughout the whole thing yielded nothing of interest.
> 
> Installed docker with
> 
> ```powershell
> Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
> Install-Package -Name Docker -ProviderName DockerMsftProvider
> ```
> 
> (which is the way MS recommends in their doc - https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/joining-windows-workers#install-docker-requires-reboot; so it's safe to assume I won't be the only one in that case)

that's interesting that not all installations of docker yield the same results. thanks for checking this @wk8